### PR TITLE
[ETCM-355] Refactor capabilities

### PR DESCRIPTION
**Pull request by lukasz-golebiewski**
# Description
Another refactoring promised in https://github.com/input-output-hk/mantis/pull/1030#discussion_r666823755
Makes unsupported capabilities unrepresentable at the type level

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockBroadcast.scala [hunk 1]
```diff
@@ -47,9 +47,10 @@ class BlockBroadcast(val etcPeerManager: ActorRef) {
     obtainRandomPeerSubset(peers.values.map(_.peer).toSet).foreach { peer =>
       val remoteStatus = peers(peer.id).peerInfo.remoteStatus
 
-      val message: MessageSerializable = remoteStatus.protocolFamily match {
-        case ETH => blockToBroadcast.as63
-        case ETC => blockToBroadcast.as64
+      val message: MessageSerializable = remoteStatus.capability match {
+        case Capability.ETH63 => blockToBroadcast.as63
+        case Capability.ETH64 => blockToBroadcast.as63
+        case Capability.ETC64 => blockToBroadcast.as64
```
Comment (dzajkowski) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockBroadcast.scala]: one more request - please rename the `as64` to `asETC64` or something similar.

File: src/main/scala/io/iohk/ethereum/network/p2p/messages/Capability.scala [hunk 2]
```diff
@@ -4,56 +4,63 @@ import io.iohk.ethereum.rlp.RLPEncodeable
...
 import io.iohk.ethereum.rlp.RLPList
 import io.iohk.ethereum.rlp.RLPSerializable
 import io.iohk.ethereum.rlp.rawDecode
 
-case class Capability(name: String, version: Byte)
+sealed trait ProtocolFamily
+object ProtocolFamily {
+  final case object ETH extends ProtocolFamily
+  final case object ETC extends ProtocolFamily
+ implicit class ProtocolFamilyEnc(val msg: ProtocolFamily) extends RLPSerializable {
```
Comment (dzajkowski) [inline src/main/scala/io/iohk/ethereum/network/p2p/messages/Capability.scala]: formatting seems off

**Review by dzajkowski — APPROVED** [review]
LGTM

File: src/main/scala/io/iohk/ethereum/network/p2p/messages/Capability.scala (hunk 2 above)
Comment (lukasz-golebiewski) [reply src/main/scala/io/iohk/ethereum/network/p2p/messages/Capability.scala]: Oops, my bad

**Comment by jvdp** [comment]
> Makes unsupported capabilities unrepresentable at the type level

Can you expand on this a bit?

Also, from the original comment:

> would be so cool to have this fail if a new block type is introduced. just a thought, no great ideas.

What does this failure look like?

**Comment by lukasz-golebiewski** [comment]
> > Makes unsupported capabilities unrepresentable at the type level
> 
> Can you expand on this a bit?
> 
> Also, from the original comment:
> 
> > would be so cool to have this fail if a new block type is introduced. just a thought, no great ideas.
> 
> What does this failure look like?

Clarified in a call

**Review by jvdp — APPROVED** [review]
👍

**Pull request merged**